pkg/runtime/handler.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+
+	pbd "istio.io/api/mixer/v1/config/descriptor"
+	"istio.io/mixer/pkg/adapter"
+	pb "istio.io/mixer/pkg/config/proto"
+	"istio.io/mixer/pkg/expr"
+	"istio.io/mixer/pkg/template"
+)
+
+type (
+	// BuilderInfoFinder is used to find specific handlers BuilderInfo for configuration.
+	BuilderInfoFinder func(name string) (*adapter.BuilderInfo, bool)
+
+	// HandlerFactory builds adapter.Handler object from adapter and instances configuration.
+	HandlerFactory interface {
+		Build(*pb.Handler, []*pb.Instance, adapter.Env) (adapter.Handler, error)
+	}
+
+	handlerFactory struct {
+		tmplRepo          template.Repository
+		typeChecker       expr.TypeChecker
+		attrDescFinder    expr.AttributeDescriptorFinder
+		builderInfoFinder BuilderInfoFinder
+
+		// protects cache
+		lock           sync.RWMutex
+		infrdTypsCache map[string]proto.Message
+	}
+
+	// Map of instance name to inferred type (proto.Message)
+	types map[string]proto.Message
+)
+
+// Create instantiates a HandlerFactory, the state of the HandlerFactory is only valid for a snapshot of a configuration.
+// Therefore, a new HandlerFactory should be created upon every configuration change.
+func Create(tmplRepo template.Repository, expr expr.TypeChecker, df expr.AttributeDescriptorFinder, builderInfoFinder BuilderInfoFinder) HandlerFactory {
+	return &handlerFactory{
+		tmplRepo:          tmplRepo,
```

Comment:
- Curious what is the reason for us to iterate over the templateInfos?
- The code you are dealing with should have the repository too, can you pass me that instead of map ? If it is tricky I can surely change it to map.. Let me know